src/Common/src/System/Runtime/InteropServices/NativeBuffer.cs

```diff
@@ -0,0 +1,151 @@
+
+using System.Runtime.CompilerServices;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Wrapper for access to the native heap. Dispose to free the memory. Try to use with using statements.
+    /// Does not allocate zero size buffers, and will free the existing native buffer if capacity is dropped to zero.
+    /// 
+    /// NativeBuffer utilizes a cache of heap buffers.
+    /// </summary>
+    /// <remarks>
+    /// Suggested use through P/Invoke: define DllImport arguments that take a byte buffer as SafeHandle.
+    /// 
+    /// Using SafeHandle will ensure that the buffer will not get collected during a P/Invoke but introduces some overhead.
+    /// (Notably AddRef and ReleaseRef will be called by the interop layer.)
+    /// 
+    /// This class is not threadsafe, changing the capacity or disposing on multiple threads risks duplicate heap
+    /// handles or worse.
+    /// </remarks>
+    internal class NativeBuffer : IDisposable
+    {
+        private static SafeHeapHandleCache s_handleCache = new SafeHeapHandleCache();
+        private static SafeHandle s_emptyHandle = new EmptySafeHandle();
+        private SafeHeapHandle _handle;
+        private ulong _capacity;
+
+        /// <summary>
+        /// Create a buffer with at least the specified initial capacity in bytes.
+        /// </summary>
+        public NativeBuffer(ulong initialMinCapacity = 0)
+        {
+            EnsureCapacity(initialMinCapacity);
+        }
+
+        protected unsafe void* VoidPointer
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                return _handle == null ? null : _handle.DangerousGetHandle().ToPointer();
+            }
+        }
+
+        protected unsafe byte* BytePointer
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                return (byte*)VoidPointer;
+            }
+        }
+
+        public static implicit operator SafeHandle(NativeBuffer buffer)
+        {
+            // Marshalling code will throw on null for SafeHandle
+            return buffer._handle ?? s_emptyHandle;
+        }
+
+        /// <summary>
+        /// The capacity of the buffer in bytes.
+        /// </summary>
+        public ulong ByteCapacity
+        {
+            get { return _capacity; }
+        }
+
+        /// <summary>
+        /// Ensure capacity in bytes is at least the given minimum.
+        /// </summary>
+        /// <exception cref="OutOfMemoryException">Thrown if unable to allocate memory when setting.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if attempting to set <paramref name="nameof(minCapacity)"/> to a value that is larger than the maximum addressable memory.</exception>
+        public virtual void EnsureCapacity(ulong minCapacity)
```

Comment:
The virtual nature of this is odd.  The derived type is overriding it and actually changing the meaning of minCapacity to mean number of characters rather than number of bytes.  If you were to give a StringBuffer out then as a NativeBuffer, someone calls this thinking they're specifying number of bytes but it actually ends up meaning number of characters.  We shouldn't overload the meaning like that.  It'd be better to have the base class have a EnsureByteCapacity method that's non virtual, and then the derived class can add an EnsureCharCapacity if it needs, or something like that.